ionic-team/stencil
# feat(@Mixin): new decorator

Much of Stencil's secret sauce comes from 'static analysis'; using the ts compiler to pull out meta from components. 
With this meta, stencil can do lots of cool stuff. Make an ultra light, lazy loading layer. Make framework specific wrappers, and allow devs to make their own wrappers!

But a reliance on static analysis does come at a cost. How could the compiler derive static meta data from components that could endlessly `extend` for example? For this reason, extending components is not allowed in Stencil.

Because I'm a stupid sadist, I decided to see if I could find a solution: A new **`@Mixin`** decorator. 
With `@Mixin` I leaned further onto static analysis in quite a naive way. Class members (and supporting statements) are simply merged together with the host, kinda like a cascade.

Usage
```
import { Component, Prop, h, ComponentInterface, Method, Mixin } from '@stencil/core';
import { Base1, Base2 } from '../bases/bases';

@Mixin(Base1)
@Mixin(Base2)
@Component(...)
export class MyInput implements ComponentInterface {
...
}
export interface MyInput extends Base1, Base2 {}
```

Notes:

- You can mixin from other stencil components or more abstract classes (those that are not components) 
- `@Mixin` order is important! - mixin members are merged like a cascade with the host. So any name clashes (variables / methods) between Base1, Base2 and the Host will be overwritten in that order.
- In order to get the most from typing, export an interface with the expected behaviour. In the case of name clashes use TS `Omit` and `Pick` e.g. `export interface MyInput extends Omit<Base1, 'offendingMethod'>, Base2 {}`
- You can further use `Omit` or `Pick` to do just that - cherry pick incoming members.
- In order to keep behaviour parseable and transparent, I have purposefully disallowed `@Mixin` nesting / recursion and trying to do so will throw an error. e.g.
```
@Mixin(Base1)
@Component(...)
export class MyInput implements ComponentInterface {
...
}

@Mixin(MyInput)
@Component(...)
export class MyNewInput implements ComponentInterface {
...
}
```
^ Error. This could be open for debate.

- No `@Component` meta is mixed in. Inluding scope, shadow, styleUrl/s and assetDirs.
- Import definitions will be brought into the host. Therefore be mindful when defining multiple mixins in a single file to not include imports that are not relevant to all mixin classes.

When writing tests I focused on 3 main use cases:

1. Mixing in from plain, generic classes.
2. Mixing in from other local stencil components.
3. Mixing in from 3rd party node_modules (it'd be nice if one day, there are defacto, community generic classes to use for example)

And that's about it. I've written quite a lot of tests for this new behaviour and all seems to be working well, but there are **bound** to be things I've missed. 

I would urge **anyone** who is remotely interested in this, please try it out and give feedback. This is quite a big change and it will need more eyes on it to have a chance at being accepted.

If this gets some traction, I will loop back over this when [source maps](https://github.com/ionic-team/stencil/pull/2892) are merged < I think they would be helpful*

*I have now made the necessary changes: mixins will work as expected with source maps

Fixes: #2844
Fixes: #2559
Fixes: #1127
Fixes: #1060
Fixes: #172

## Pull request checklist
 Please check if your PR fulfills the following requirements:
 
    * They've been added ^ here
 * [x]  Build (`npm run build`) was run locally 
 
 ## Pull request type
 Please check the type of change your PR introduces:
 
 
 ## What is the current behavior?
 There is no easy way to extend from other classes.
 
 ## What is the new behavior?
 As described above.
 
 ## Does this introduce a breaking change?
 
 ## Manual Testing
 * Ran `npm build` and `npm link` from the root of the repo directory (on this branch)
 * Spun up a boilerplate Stencil app using the [Stencil CLI](https://stenciljs.com/docs/getting-started) `npm init stencil`
 * Linked my project with the instructions found in the [Contributing guidelines](https://github.com/ionic-team/stencil/blob/master/.github/CONTRIBUTING.md#testing-changes-against-a-project-locally)
 * Added multiple `@Mixin` decorated components, with multiple diffrent type of mixin classes.
 
 ```shell
 npm run start
 ```
 * All functions and builds as expected. e.g. Changing / saving Mixed in classes cause `@Mixin` components to reload 👍 
```shell
 npm run build
 ```
 * All functions and builds as expected. Built components have all relevant mixed in class members present 👍
